### PR TITLE
DBranch

### PR DESCRIPTION
**Pull request by nguyendan1337**
- Added Q.py file with is_terminal_state() and get_next_action()
- bug fixes for BFS.py
- began main function

### CODE REVIEW

File: Q.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+# translation = {39: None}
+# print(str(board).translate(translation))
+# boxes = BFS.get_reachable_boxes(board, agent)
+# pprint.pprint(boxes)
+
+#TEST GET NEXT ACTION
+# q_table = {(4, 7): {'Up': 1, 'Down': 2, 'Left': 3, 'Right': 4}}
+# q_table = {}
+# action = get_next_action(boxes, .9, q_table)
+# print(action)
```
Comment (brookekelseyryan) [inline Q.py]: move this to a separate Q_test.py file

File: BFS.py [hunk 2]
```diff
@@ -110,18 +112,19 @@ def expand(board, node, boxes, path):
...
+# boxes = get_reachable_boxes(sokoban.board, sokoban.agent_location)
 # pprint.pprint(boxes)
 
 # FROM TEXT LEVELS
-b = b.Board("test/input/levels/level0.txt")
+# np.set_printoptions(edgeitems=30, linewidth=100000,formatter=dict(float=lambda x: "%.3g" % x))
+# b = b.Board("test/input/levels/level0.txt")
 # b = b.Board("test/input/levels/level1.txt")
 # b = b.Board("test/input/levels/level2.txt")
 # b = b.Board("test/input/levels/level47.txt")
```
Comment (brookekelseyryan) [inline BFS.py]: cleanup or move these comments to a test file

File: BFS.py [hunk 3]
```diff
@@ -110,18 +112,19 @@ def expand(board, node, boxes, path):
...
-agent = sb.Agent(b.agentX, b.agentY)
-boxes = getReachableBoxes(board, agent)
-pprint.pprint(boxes)
+# board = b.board
+# agent = b.agent
+# translation = {39: None}
+# print(str(board).translate(translation))
+# boxes = get_reachable_boxes(board, agent)
+# pprint.pprint(boxes)
+
```
Comment (brookekelseyryan) [inline BFS.py]: cleanup or move to a test file

File: BFS.py [hunk 4]
```diff
@@ -110,18 +112,19 @@ def expand(board, node, boxes, path):
 # sokoban = sb.Sokoban("test/input/sokoban/sokoban00.txt")
 # sokoban = sb.Sokoban("test/input/sokoban/sokoban01.txt")
 # sokoban.print()
-# agent = sb.Agent(sokoban.agentX, sokoban.agentY)
-# boxes = getReachableBoxes(sokoban.board, agent)
+# boxes = get_reachable_boxes(sokoban.board, sokoban.agent_location)
 # pprint.pprint(boxes)
```
Comment (brookekelseyryan) [inline BFS.py]: cleanup or move to test file

File: BFS.py [hunk 5]
```diff
@@ -3,8 +3,6 @@
 import queue as q
 import board as b
 import pprint
-np.set_printoptions(edgeitems=30, linewidth=100000,formatter=dict(float=lambda x: "%.3g" % x))
```
Comment (brookekelseyryan) [inline BFS.py:6]: are these print options specific to the BFS class only?  Or do we want this in the main.py file?

File: Q.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    # choose a random box and a random move from the reachable boxes
+    box = random.choice(list(boxes.items()))
+    move = random.choice(list(box[1].keys()))
+    box_move = {box[0]: move}
+
+    # if a randomly chosen value between 0 and 1 is less than epsilon,
+    # then of the reachable boxes, choose box with the move with the highest q value
+    # else return the random box and move
+    r = np.random.random()
+    # print(r)
```
Comment (brookekelseyryan) [inline Q.py]: remove?

File: Q.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+RIGHT = "Right"
+
+# determine if any boxes are terminal states
+def is_terminal_state(board, boxes, rewards):
+
+    #if all boxes are in goals, terminal will remain true
+    for box in boxes:
+        # if the reward for this box's location is -100, then it is in a corner terminal state
+        if rewards[box[0], box[1]] == -100:
+            return True, DEAD
```
Comment (brookekelseyryan) [inline Q.py:22]: I like this pattern

File: Q.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+LEFT = "Left"
+RIGHT = "Right"
+
+# determine if any boxes are terminal states
+def is_terminal_state(board, boxes, rewards):
+
+    #if all boxes are in goals, terminal will remain true
+    for box in boxes:
+        # if the reward for this box's location is -100, then it is in a corner terminal state
+        if rewards[box[0], box[1]] == -100:
```
Comment (brookekelseyryan) [inline Q.py:21]: is it possible we could get an out of bounds exception?  are we enforcing somewhere that the boxes parameter will always have tuple pairs of boxes?

File: BFS.py [hunk 9]
```diff
@@ -52,52 +50,56 @@ def expand(board, node, boxes, path):
     ##UP##
     if board[node[0] - 1][node[1]] != WALL:                         #if UP is not a wall
         if board[node[0] - 1][node[1]] == BOX:                          #if UP is a box
-            if board[node[0] - 2][node[1]] != WALL:                         #if you can push the box UP
+            if board[node[0] - 2][node[1]] != WALL\
```
Comment (brookekelseyryan) [inline BFS.py:49]: is an array index out of bounds exception ever possible in this logic?

File: main.py [hunk 10]
```diff
@@ -1,7 +1,54 @@
...
+sokoban = sb.Sokoban("test/input/sokoban/sokoban01.txt")
+sokoban.print()
+
+# define training parameters
+epsilon = 0.9  # the percentage of time when we should take the best action (instead of a random action)
+discount_factor = 0.9  # discount factor for future rewards
+learning_rate = 0.9  # the rate at which the AI agent should learn
+
+# run through 1000 training episodes
+for episode in range(1000):
```
Comment (brookekelseyryan) [inline main.py]: move 1000 to a constant

File: main.py [hunk 11]
```diff
@@ -1,7 +1,54 @@
...
+
+        # choose which box and move to make
+        action = Q.get_next_action(boxes, q_table, epsilon)
+
+        #perform the action
+        #update agent and box locations
+        #append path to history
+
+        #update Q values in Q Table
+        # q_table = Q.update_q_table(q_table, rewards)
```
Comment (brookekelseyryan) [inline main.py:52]: commented out code--cleanup

**Review by brookekelseyryan — CHANGES_REQUESTED** [review]

File: BFS.py (hunk 5 above)
Comment (nguyendan1337) [reply BFS.py:6]: just for nice output of board when testing

File: BFS.py (hunk 9 above)
Comment (nguyendan1337) [reply BFS.py:49]: nope

File: Q.py (hunk 8 above)
Comment (nguyendan1337) [reply Q.py:21]: these are the box's x and y coordinates

File: Q.py (hunk 7 above)
Comment (nguyendan1337) [reply Q.py:22]: lol

File: main.py (hunk 11 above)
Comment (nguyendan1337) [reply main.py:52]: it will be done

File: Q.py (hunk 8 above)
Comment (brookekelseyryan) [reply Q.py:21]: I still think this needs to be renamed.  one possible suggestion is 
for **coordinate** in boxes 

also add a comment above the check so we will be able to understand it when referencing later

**Review by brookekelseyryan — COMMENTED** [review]
fix the for box in boxes naming and add comments above for clarity

**Pull request merged**